Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift

```diff
@@ -0,0 +1,191 @@
+
+    let requestBody: HTTPBody?
+    var hasAlreadyIteratedRequestBody: Bool
+    var hasSuspendedURLSessionTask: Bool
+    let requestStreamBufferSize: Int
+    var requestStream: HTTPBodyOutputStreamBridge?
+
+    typealias ResponseContinuation = CheckedContinuation<URLResponse, any Error>
+    var responseContinuation: ResponseContinuation?
+
+    typealias ResponseBodyStream = AsyncBackpressuredStream<HTTPBody.ByteChunk, any Error>
+    var responseBodyStream: ResponseBodyStream
+    var responseBodyStreamSource: ResponseBodyStream.Source
+
+    /// This lock is taken for the duration of all delegate callbacks to protect the mutable delegate state.
+    ///
+    /// Although all the delegate callbacks are performed on the session's `delegateQueue`, there is no guarantee that
+    /// this is a _serial_ queue.
+    ///
+    /// Regardless of the type of delegate queue, URLSession will attempt to order the callbacks for each task in a
+    /// sensible way, but it cannot be guaranteed, specifically when the URLSession task is cancelled.
+    ///
+    /// Therefore, even though the `suspend()`, `resume()`, and `cancel()` URLSession methods are thread-safe, we need
+    /// to protect any mutable state within the delegate itself.
+    let callbackLock = NSLock()
+
+    /// In addition to the callback lock, there is one point of rentrancy, where the response stream callback gets fired
+    /// immediately, for this we have a different lock, which protects `hasSuspendedURLSessionTask`.
+    let hasSuspendedURLSessionTaskLock = NSLock()
+
+    /// Use `bidirectionalStreamingRequest(for:baseURL:requestBody:requestStreamBufferSize:responseStreamWatermarks:)`.
+    init(requestBody: HTTPBody?, requestStreamBufferSize: Int, responseStreamWatermarks: (low: Int, high: Int)) {
+        self.requestBody = requestBody
+        self.hasAlreadyIteratedRequestBody = false
+        self.hasSuspendedURLSessionTask = false
+        self.requestStreamBufferSize = requestStreamBufferSize
+        (self.responseBodyStream, self.responseBodyStreamSource) = AsyncBackpressuredStream.makeStream(
+            backPressureStrategy: .highLowWatermarkWithElementCounts(
+                lowWatermark: responseStreamWatermarks.low,
+                highWatermark: responseStreamWatermarks.high
+            )
+        )
+    }
+
+    func urlSession(_ session: URLSession, needNewBodyStreamForTask task: URLSessionTask) async -> InputStream? {
+        callbackLock.withLock {
+            debug("Task delegate: needNewBodyStreamForTask")
+            // If the HTTP body cannot be iterated multiple times then bad luck; the only thing
+            // we can do is cancel the task and return nil.
+            if hasAlreadyIteratedRequestBody {
+                guard requestBody!.iterationBehavior == .multiple else {
+                    debug("Task delegate: Cannot rewind request body, cancelling task")
+                    task.cancel()
```

Comment:
Returning nil is sufficient, no need to do both

Reply:
Hm, I just tried removing `task.cancel()` here and `testHTTPRedirect_singleIterationBehavior_fails` never completes.

Without the `task.cancel()`, the `needNewBodyStreamForTask` callback is called in a loop.